File: experimentos/consola-multilenguaje/gb-hook.php
<?php
/**
 * gb-hook.php — PHP crash capture hook for galaxy-brain (schema v2)
 *
 * Install: php.ini → auto_prepend_file = /path/to/gb-hook.php
 *
 * Triple handler:
 *   1. set_exception_handler  — uncaught exceptions
 *   2. set_error_handler      — runtime errors (E_WARNING, E_NOTICE, etc.)
 *   3. register_shutdown_function + error_get_last() — fatal errors
 *
 * Zero external dependencies. Dormant until a crash occurs.
 * Appends a JSON record to ~/.galaxy-brain/crashes.jsonl.
 */

(function () {
    // Resolve home directory portably.
    $home = getenv('HOME') ?: getenv('USERPROFILE') ?: (
        getenv('HOMEDRIVE') && getenv('HOMEPATH')
            ? getenv('HOMEDRIVE') . getenv('HOMEPATH')
            : sys_get_temp_dir()
    );

    $crashesDir  = $home . DIRECTORY_SEPARATOR . '.galaxy-brain';
    $crashesFile = $crashesDir . DIRECTORY_SEPARATOR . 'crashes.jsonl';

    // ---------------------------------------------------------------
    // Helpers (closures to avoid polluting global namespace)
    // ---------------------------------------------------------------

    /**
     * Walk up from $dir looking for .git.
     */
    $findProjectRoot = function (string $dir): ?string {
        $cur = realpath($dir) ?: $dir;
        while (true) {
            if (file_exists($cur . DIRECTORY_SEPARATOR . '.git')) {
                return $cur;
            }
            $parent = dirname($cur);
            if ($parent === $cur) {
                return null;  // filesystem root
            }
            $cur = $parent;
        }
    };

    /**
     * Redact argv: keep flag names, replace values with <val>.
     */
    $redactArgv = function (array $argv): array {
        $result = [];
        $count  = count($argv);
        for ($i = 0; $i < $count; $i++) {
            $arg = $argv[$i];
            if (preg_match('/^--?[a-zA-Z]/', $arg)) {
                $eqPos = strpos($arg, '=');
                if ($eqPos !== false) {
                    $result[] = substr($arg, 0, $eqPos + 1) . '<val>';
                } else {
                    $result[] = $arg;
                    if ($i + 1 < $count && !preg_match('/^--?[a-zA-Z]/', $argv[$i + 1])) {
                        $result[] = '<val>';
                        $i++;
                    }
                }
            } else {
                $result[] = $arg;
            }
        }
        return $result;
    };

    /**
     * Parse a backtrace array (from debug_backtrace or Exception::getTrace)
     * into schema-v2 frames.
     */
    $parseTrace = function (array $trace): array {
        $frames = [];
        foreach ($trace as $frame) {
            $frames[] = [
                'file'     => $frame['file']     ?? '<internal>',
                'line'     => $frame['line']      ?? 0,
                'column'   => 0,  // PHP doesn't provide column info
                'function' => isset($frame['class'])
                    ? $frame['class'] . ($frame['type'] ?? '::') . ($frame['function'] ?? '')
                    : ($frame['function'] ?? '<unknown>'),
            ];
        }
        return $frames;
    };

    /**
     * Build a schema-v2 crash record.
     */
    $buildRecord = function (
        string $type,
        string $message,
        string $origin,
        array  $frames,
        string $rawTrace
    ) use ($findProjectRoot, $redactArgv): array {
        $cwd = getcwd() ?: '';
        $sessionId = getenv('GB_SESSION_ID') ?: 'unknown';
        // PHP doesn't have a built-in ppid function; use posix_getppid if available.
        $ppid = function_exists('posix_getppid') ? posix_getppid() : null;
        return [
            'schema'     => 2,
            'ts'         => date('c'),  // ISO 8601 with timezone
            'session_id' => $sessionId,
            'language'   => 'php',
            'exception'  => [
                'type'    => $type,
                'message' => $message,
                'origin'  => $origin,
            ],
            'frames'    => $frames,
            'process'   => [
                'cwd'        => $cwd,
                'project'    => $findProjectRoot($cwd),
                'argv_forma' => $redactArgv($_SERVER['argv'] ?? []),
                'runtime'    => 'php ' . PHP_VERSION,
                'pid'        => getmypid(),
                'ppid'       => $ppid,
            ],
            'traceback'      => $rawTrace,
            'capture_method' => 'hook',
        ];
    };

    /**
     * Append record to crashes.jsonl. Swallows all errors.
     */
    $writeRecord = function (array $record) use ($crashesDir, $crashesFile): void {
        try {
            if (!is_dir($crashesDir)) {
                @mkdir($crashesDir, 0755, true);
            }
            @file_put_contents(
                $crashesFile,
                json_encode($record, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n",
                FILE_APPEND | LOCK_EX
            );
        } catch (\Throwable $e) {
            // Silent fail.
        }
    };

    // Track whether we've already captured (avoid double-capture from
    // exception handler + shutdown function for the same error).
    $captured = false;

    // ---------------------------------------------------------------
    // 1. Uncaught exception handler
    // ---------------------------------------------------------------
    $previousExceptionHandler = set_exception_handler(
        function (\Throwable $ex) use (
            &$captured, $buildRecord, $parseTrace, $writeRecord,
            &$previousExceptionHandler
        ): void {
            if (!$captured) {
                $captured = true;
                try {
                    $frames   = $parseTrace($ex->getTrace());
                    $rawTrace = $ex->getTraceAsString();
                    $record   = $buildRecord(
                        get_class($ex),
                        $ex->getMessage(),
                        'exception',
                        $frames,
                        $rawTrace
                    );
                    $writeRecord($record);
                } catch (\Throwable $e) {
                    // Silent fail.
                }
            }

            // Chain to any previously installed handler.
            if (is_callable($previousExceptionHandler)) {
                ($previousExceptionHandler)($ex);
            }
        }
    );

    // ---------------------------------------------------------------
    // 2. Error handler (non-fatal errors: warnings, notices, etc.)
    //    Only capture errors that would be fatal or are severe.
    // ---------------------------------------------------------------
    $fatalErrorTypes = E_ERROR | E_PARSE | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR;
    $previousErrorHandler = set_error_handler(
        function (
            int $errno,
            string $errstr,
            string $errfile = '',
            int $errline = 0
        ) use (
            &$captured, $fatalErrorTypes,
            $buildRecord, $parseTrace, $writeRecord,
            &$previousErrorHandler
        ): bool {
            // Only capture error types that are severe enough.
            if ($errno & $fatalErrorTypes) {
                if (!$captured) {
                    $captured = true;
                    try {
                        $frames   = $parseTrace(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS));
                        $rawTrace = "Error [{$errno}] in {$errfile}:{$errline} — {$errstr}";
                        $record   = $buildRecord(
                            'E_' . $errno,
                            $errstr,
                            'error_handler',
                            $frames,
                            $rawTrace
                        );
                        $writeRecord($record);
                    } catch (\Throwable $e) {
                        // Silent fail.
                    }
                }
            }

            // Chain to previous handler or let PHP's default handle it.
            if (is_callable($previousErrorHandler)) {
                return ($previousErrorHandler)($errno, $errstr, $errfile, $errline);
            }
            return false;  // Let PHP's default error handler run.
        }
    );

    // ---------------------------------------------------------------
    // 3. Shutdown function — catches fatal errors that bypass the
    //    error handler (E_ERROR, E_PARSE, segfaults reported via
    //    error_get_last).
    // ---------------------------------------------------------------
    register_shutdown_function(
        function () use (
            &$captured, $fatalErrorTypes,
            $buildRecord, $writeRecord
        ): void {
            $error = error_get_last();
            if ($error === null) {
                return;
            }
            // Only capture fatal-class errors.
            if (!($error['type'] & $fatalErrorTypes)) {
                return;
            }
            if ($captured) {
                return;
            }
            $captured = true;
            try {
                $frames = [[
                    'file'     => $error['file'] ?? '<unknown>',
                    'line'     => $error['line'] ?? 0,
                    'column'   => 0,
                    'function' => '<fatal>',
                ]];
                $rawTrace = "Fatal [{$error['type']}] in {$error['file']}:{$error['line']} — {$error['message']}";
                $record   = $buildRecord(
                    'E_FATAL_' . $error['type'],
                    $error['message'],
                    'shutdown',
                    $frames,
                    $rawTrace
                );
                $writeRecord($record);
            } catch (\Throwable $e) {
                // Silent fail.
            }
        }
    );
})();

File: experimentos/consola-multilenguaje/gb-hook.lua
#!/usr/bin/env lua
--- Galaxy Brain crash capture launcher for Lua.
---
--- Wraps the user's script in xpcall, captures crash details (locals,
--- stack frames) via debug.getlocal/debug.getinfo, and writes a JSON
--- schema v2 record to ~/.galaxy-brain/crashes.jsonl before re-raising.
---
--- Usage:
---   lua gb-hook.lua user_script.lua [args...]

-- ===================================================================
-- Minimal JSON encoder (no external deps)
-- ===================================================================

local function json_escape(s)
    s = s:gsub('\\', '\\\\')
    s = s:gsub('"', '\\"')
    s = s:gsub('\n', '\\n')
    s = s:gsub('\r', '\\r')
    s = s:gsub('\t', '\\t')
    s = s:gsub('[\x00-\x1f]', function(c)
        return string.format('\\u%04x', c:byte())
    end)
    return s
end

local json_encode  -- forward declaration for recursion

local function json_encode_value(val, depth)
    depth = depth or 0
    if depth > 20 then return '"<max depth>"' end

    local t = type(val)
    if val == nil then
        return 'null'
    elseif t == 'boolean' then
        return val and 'true' or 'false'
    elseif t == 'number' then
        if val ~= val then return '"NaN"' end         -- NaN
        if val == math.huge then return '"Inf"' end
        if val == -math.huge then return '"-Inf"' end
        return tostring(val)
    elseif t == 'string' then
        return '"' .. json_escape(val) .. '"'
    elseif t == 'table' then
        return json_encode(val, depth + 1)
    else
        return '"' .. json_escape(tostring(val)) .. '"'
    end
end

json_encode = function(tbl, depth)
    depth = depth or 0
    if depth > 20 then return '"<max depth>"' end

    -- Detect array vs object: a table is an array if keys are 1..#tbl.
    local is_array = true
    local n = #tbl
    if n == 0 then
        -- could be empty array or empty object — treat as object if any key exists
        if next(tbl) ~= nil then is_array = false end
    else
        for k, _ in pairs(tbl) do
            if type(k) ~= 'number' or k < 1 or k > n or k ~= math.floor(k) then
                is_array = false
                break
            end
        end
    end

    local parts = {}
    if is_array then
        for i = 1, n do
            parts[#parts + 1] = json_encode_value(tbl[i], depth)
        end
        return '[' .. table.concat(parts, ',') .. ']'
    else
        for k, v in pairs(tbl) do
            local key = type(k) == 'string' and k or tostring(k)
            parts[#parts + 1] = '"' .. json_escape(key) .. '":' .. json_encode_value(v, depth)
        end
        return '{' .. table.concat(parts, ',') .. '}'
    end
end

-- ===================================================================
-- Helpers
-- ===================================================================

local function get_home()
    return os.getenv('HOME') or os.getenv('USERPROFILE') or '.'
end

local function path_join(...)
    local sep = package.config:sub(1, 1)  -- '/' or '\'
    return table.concat({...}, sep)
end

local function file_exists(path)
    local f = io.open(path, 'r')
    if f then f:close(); return true end
    return false
end

local function detect_project_root()
    -- Walk up from the working directory looking for .git
    local sep = package.config:sub(1, 1)
    -- Use io.popen to get cwd portably
    local handle = io.popen('cd')  -- Windows
    if not handle then
        handle = io.popen('pwd')   -- Unix
    end
    if not handle then return nil end
    local cwd = handle:read('*l')
    handle:close()
    if not cwd or cwd == '' then return nil end

    -- Normalise to forward slashes for easier manipulation.
    cwd = cwd:gsub('\\', '/')

    while true do
        local git_path = cwd .. '/.git'
        if file_exists(git_path) then
            return cwd
        end
        local parent = cwd:match('^(.+)/[^/]+$')
        if not parent then return nil end
        cwd = parent
    end
end

local function iso8601_now()
    return os.date('!%Y-%m-%dT%H:%M:%SZ')
end

local function mkdir_p(path)
    -- Best-effort recursive mkdir.
    local sep = package.config:sub(1, 1)
    if sep == '\\' then
        os.execute('mkdir "' .. path:gsub('/', '\\') .. '" 2>nul')
    else
        os.execute('mkdir -p "' .. path .. '" 2>/dev/null')
    end
end

-- ===================================================================
-- Crash capture
-- ===================================================================

local function capture_locals(level)
    local locals = {}
    local i = 1
    while true do
        local name, value = debug.getlocal(level, i)
        if not name then break end
        if name:sub(1, 1) ~= '(' then  -- skip internal vars like "(*temporary)")
            locals[name] = tostring(value)
        end
        i = i + 1
    end
    return locals
end

local function capture_stack(level)
    local frames = {}
    local i = level
    while true do
        local info = debug.getinfo(i, 'nSlf')
        if not info then break end
        frames[#frames + 1] = {
            source     = info.source or '?',
            short_src  = info.short_src or '?',
            what       = info.what or '?',
            name       = info.name,
            line       = info.currentline,
            linedefined = info.linedefined,
        }
        i = i + 1
        if #frames > 50 then break end  -- safety cap
    end
    return frames
end

local function write_crash(err_msg, locals, stack_frames)
    -- Build schema-v2 compliant frames from raw stack
    local v2_frames = {}
    for _, f in ipairs(stack_frames) do
        local src = f.source or '?'
        -- Strip leading '@' that Lua adds to file sources
        if src:sub(1, 1) == '@' then src = src:sub(2) end
        v2_frames[#v2_frames + 1] = {
            file     = src,
            line     = f.line or 0,
            column   = 0,
            ['function'] = f.name or f.what or '?',
            locals   = nil,  -- filled below for the crash frame
        }
    end
    -- Attach locals to the first user frame (the one that errored)
    if #v2_frames > 0 and locals and next(locals) then
        v2_frames[1].locals = locals
    end

    local cwd_handle = io.popen(package.config:sub(1,1) == '\\' and 'cd' or 'pwd')
    local cwd = cwd_handle and cwd_handle:read('*l') or '.'
    if cwd_handle then cwd_handle:close() end

    local session_id = os.getenv('GB_SESSION_ID') or 'unknown'
    -- Lua has no portable ppid; try GB_PPID env var set by gb-run.py
    local ppid_str = os.getenv('GB_PPID')
    local ppid_val = ppid_str and tonumber(ppid_str) or nil

    local record = {
        schema  = 2,
        ts      = iso8601_now(),
        session_id = session_id,
        language = 'lua',
        exception = {
            type    = 'runtime_error',
            message = tostring(err_msg),
            origin  = 'xpcall',
        },
        frames  = v2_frames,
        process = {
            cwd     = cwd,
            project = detect_project_root(),
            runtime = _VERSION,
            pid     = nil,
            ppid    = ppid_val,
        },
        traceback = debug.traceback(tostring(err_msg), 2),
        capture_method = 'hook',
    }

    local home = get_home()
    local dir  = path_join(home, '.galaxy-brain')
    local file = path_join(dir, 'crashes.jsonl')

    mkdir_p(dir)

    local fh = io.open(file, 'a')
    if fh then
        fh:write(json_encode(record) .. '\n')
        fh:close()
    else
        io.stderr:write('[gb-hook.lua] WARNING: could not write crash record to ' .. file .. '\n')
    end
end

-- ===================================================================
-- Error handler (runs before stack unwinds)
-- ===================================================================

local captured_locals = nil
local captured_stack  = nil

local function error_handler(err)
    -- Level 2 = the frame that errored (1 = this handler, 2 = caller).
    captured_locals = capture_locals(2)
    captured_stack  = capture_stack(2)
    write_crash(err, captured_locals, captured_stack)
    return err  -- return original error for re-raise
end

-- ===================================================================
-- Main: load and run the user script
-- ===================================================================

local function main()
    local script = arg[1]
    if not script then
        io.stderr:write('Usage: lua gb-hook.lua <script.lua> [args...]\n')
        os.exit(1)
    end

    -- Shift args so the user script sees its own args as arg[1], arg[2], ...
    local new_arg = { [0] = script }
    for i = 2, #arg do
        new_arg[i - 1] = arg[i]
    end
    arg = new_arg

    local chunk, load_err = loadfile(script)
    if not chunk then
        io.stderr:write('gb-hook.lua: failed to load script: ' .. tostring(load_err) .. '\n')
        os.exit(1)
    end

    local ok, err = xpcall(chunk, error_handler)
    if not ok then
        -- Re-raise so the original exit behaviour is preserved.
        error(err, 0)
    end
end

main()
